src/main/java/de/tum/in/www1/artemis/service/ProgressBasedAchievementService.java

```diff
@@ -21,12 +25,57 @@
 
     private final static long EXERCISES_AMOUNT_BRONZE = 5L;
 
+    private final static long EXERCISES_AMOUNT_UNRANKED = 1L;
+
     private final static long MIN_SCORE_TO_QUALIFY = 50L;
 
-    public ProgressBasedAchievementService(AchievementRepository achievementRepository) {
+    public ProgressBasedAchievementService(StudentParticipationRepository studentParticipationRepository, AchievementRepository achievementRepository) {
+        this.studentParticipationRepository = studentParticipationRepository;
         this.achievementRepository = achievementRepository;
     }
 
+    /**
+     * Checks whether a user earned progress based achievements
+     * @param result the currently saved result
+     * @param course the given course
+     * @param user the given user
+     * @param achievements all progress based achievements within the given course
+     * @return the highest rank reached, returns null if no rank was reached
+     */
+    public AchievementRank checkForAchievement(Result result, Course course, User user, Set<Achievement> achievements) {
+        if (!achievements.iterator().hasNext()) {
+            return null;
+        }
+        var minScore = achievements.iterator().next().getMinScoreToQualify();
+        var participations = studentParticipationRepository.findAllByCourseIdAndUserId(course.getId(), user.getId());
+        var numberOfExercises = 0;
+        if (result.getScore() != null && result.getScore() >= minScore) {
+            numberOfExercises++;
+        }
+        for (var participation : participations) {
+            var latestResult = participation.findLatestResult();
+            if (latestResult == null || result.getParticipation() == participation) {
```

Comment:
No sure if the strict equality would hold. Better check for the participation id instead.

Reply:
oh right, good point